File: backend/src/services/audioProbe.ts
import { execFile } from 'child_process';
import { promisify } from 'util';
import fs from 'fs/promises';
import path from 'path';
import os from 'os';

const execFileAsync = promisify(execFile);
const exec = (file: string, args: string[]) =>
  execFileAsync(file, args, { maxBuffer: 256 * 1024 * 1024 });

// Concatenate the audio files listed in a concat-demuxer manifest into one mp3,
// optionally scaling volume. Always decodes and re-encodes: raw mp3 byte-concat
// drops ~20ms of audio at every segment boundary (frames aren't independently
// joinable), which desyncs the read-along timeline; decoding to PCM first makes
// the join sample-accurate so segment durations sum exactly to the output.
export async function concatAudio(listPath: string, output: string, volume = 1): Promise<void> {
  const args = ['-y', '-v', 'error', '-f', 'concat', '-safe', '0', '-i', listPath];
  if (volume !== 1) args.push('-filter:a', `volume=${volume}`);
  args.push('-c:a', 'libmp3lame', '-q:a', '2', '-write_xing', '1', output);
  await exec('ffmpeg', args);
}

// Re-encode an in-memory mp3 with a volume scale (used to boost title segments
// before they're concatenated into the chapter).
export async function applyVolume(buffer: Buffer, volume: number): Promise<Buffer> {
  if (volume === 1) return buffer;
  const base = path.join(os.tmpdir(), `vol_${process.pid}_${Date.now()}_${Math.random().toString(36).slice(2)}`);
  const inPath = `${base}_in.mp3`;
  const outPath = `${base}_out.mp3`;
  try {
    await fs.writeFile(inPath, buffer);
    await exec('ffmpeg', [
      '-y', '-v', 'error', '-i', inPath,
      '-filter:a', `volume=${volume}`, '-c:a', 'libmp3lame', '-q:a', '2',
      outPath,
    ]);
    return await fs.readFile(outPath);
  } finally {
    await fs.unlink(inPath).catch(() => {});
    await fs.unlink(outPath).catch(() => {});
  }
}

// Sample rate and channel count of a file's first audio stream, so generated
// silence matches the TTS segments it sits between.
export async function probeAudioFormat(file: string): Promise<{ sampleRate: number; channels: number }> {
  const { stdout } = await exec('ffprobe', [
    '-v', 'error',
    '-select_streams', 'a:0',
    '-show_entries', 'stream=sample_rate,channels',
    '-of', 'default=noprint_wrappers=1',
    file,
  ]);
  let sampleRate = 0;
  let channels = 0;
  for (const line of stdout.trim().split('\n')) {
    const [key, value] = line.split('=');
    if (key === 'sample_rate') sampleRate = parseInt(value, 10);
    else if (key === 'channels') channels = parseInt(value, 10);
  }
  return { sampleRate: sampleRate || 24000, channels: channels || 1 };
}

// An mp3 buffer of pure silence, format-matched so it can be raw-concatenated
// with TTS segments before the final re-encode.
export async function generateSilence(durationSecs: number, sampleRate: number, channels: number): Promise<Buffer> {
  const tmp = path.join(os.tmpdir(), `silence_${process.pid}_${Date.now()}_${Math.random().toString(36).slice(2)}.mp3`);
  try {
    await exec('ffmpeg', [
      '-y', '-v', 'error',
      '-f', 'lavfi',
      '-i', `anullsrc=r=${sampleRate}:cl=${channels === 1 ? 'mono' : 'stereo'}`,
      '-t', String(durationSecs),
      '-c:a', 'libmp3lame', '-q:a', '2',
      tmp,
    ]);
    return await fs.readFile(tmp);
  } finally {
    await fs.unlink(tmp).catch(() => {});
  }
}

// True decoded duration in seconds: the number of samples the decoder actually
// emits, which is what concat and playback use. This differs from format=duration
// for files carrying a LAME/Xing gapless tag (libmp3lame output — generated
// silence, re-encoded title boosts), where format=duration counts the encoder
// delay/padding the decoder strips. Summed decoded durations equal the concat
// output exactly, so the read-along timeline stays locked to the audio.
export async function decodedDurationSecs(file: string): Promise<number> {
  const { stdout } = await exec('ffmpeg', ['-v', 'error', '-i', file, '-f', 'null', '-progress', 'pipe:1', '-']);
  const matches = stdout.match(/out_time_us=(\d+)|out_time_ms=(\d+)/g) ?? [];
  const last = matches[matches.length - 1];
  const us = last ? parseInt(last.split('=')[1], 10) : NaN;
  if (!isFinite(us) || us <= 0) throw new Error(`ffmpeg returned no duration for ${file}`);
  return us / 1e6;
}

export async function probeDurationSecs(file: string): Promise<number> {
  const { stdout } = await exec('ffprobe', [
    '-v', 'error',
    '-show_entries', 'format=duration',
    '-of', 'default=noprint_wrappers=1:nokey=1',
    file,
  ]);
  const secs = parseFloat(stdout.trim());
  if (!isFinite(secs) || secs <= 0) throw new Error(`ffprobe returned no duration for ${file}`);
  return secs;
}

// Duration of an in-memory mp3 (for engines that don't return a duration header).
export async function probeMp3Buffer(buffer: Buffer): Promise<number> {
  const tmp = path.join(os.tmpdir(), `probe_${process.pid}_${Date.now()}_${Math.random().toString(36).slice(2)}.mp3`);
  try {
    await fs.writeFile(tmp, buffer);
    return await probeDurationSecs(tmp);
  } finally {
    await fs.unlink(tmp).catch(() => {});
  }
}

File: backend/internal/svc/audioprobe/audioprobe.go
package audioprobe

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

func run(name string, args ...string) (string, error) {
	out, err := exec.Command(name, args...).Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok && len(ee.Stderr) > 0 {
			return "", fmt.Errorf("%s: %s", name, strings.TrimSpace(string(ee.Stderr)))
		}
		return "", err
	}
	return string(out), nil
}

func tmpFile(prefix string) string {
	f, err := os.CreateTemp("", prefix+"*.mp3")
	if err != nil {
		return filepath.Join(os.TempDir(), prefix+strconv.Itoa(os.Getpid())+".mp3")
	}
	name := f.Name()
	f.Close()
	return name
}

// ConcatAudio concatenates the audio files listed in a concat-demuxer manifest
// into one mp3, optionally scaling volume. Always decodes and re-encodes: raw
// mp3 byte-concat drops ~20ms at every segment boundary, which desyncs the
// read-along timeline; decoding to PCM first makes the join sample-accurate.
func ConcatAudio(listPath, output string, volume float64) error {
	args := []string{"-y", "-v", "error", "-f", "concat", "-safe", "0", "-i", listPath}
	if volume != 1 {
		args = append(args, "-filter:a", fmt.Sprintf("volume=%g", volume))
	}
	args = append(args, "-c:a", "libmp3lame", "-q:a", "2", "-write_xing", "1", output)
	_, err := run("ffmpeg", args...)
	return err
}

// ApplyVolume re-encodes an in-memory mp3 with a volume scale (used to boost
// title segments before chapter concat).
func ApplyVolume(buffer []byte, volume float64) ([]byte, error) {
	if volume == 1 {
		return buffer, nil
	}
	inPath := tmpFile("vol_in_")
	outPath := tmpFile("vol_out_")
	defer os.Remove(inPath)
	defer os.Remove(outPath)
	if err := os.WriteFile(inPath, buffer, 0o644); err != nil {
		return nil, err
	}
	if _, err := run("ffmpeg",
		"-y", "-v", "error", "-i", inPath,
		"-filter:a", fmt.Sprintf("volume=%g", volume), "-c:a", "libmp3lame", "-q:a", "2",
		outPath); err != nil {
		return nil, err
	}
	return os.ReadFile(outPath)
}

// ProbeAudioFormat reports sample rate and channel count of a file's first
// audio stream, so generated silence matches the TTS segments around it.
func ProbeAudioFormat(file string) (sampleRate, channels int, err error) {
	out, err := run("ffprobe",
		"-v", "error",
		"-select_streams", "a:0",
		"-show_entries", "stream=sample_rate,channels",
		"-of", "default=noprint_wrappers=1",
		file)
	if err != nil {
		return 0, 0, err
	}
	for _, line := range strings.Split(strings.TrimSpace(out), "\n") {
		key, value, _ := strings.Cut(line, "=")
		switch key {
		case "sample_rate":
			sampleRate, _ = strconv.Atoi(value)
		case "channels":
			channels, _ = strconv.Atoi(value)
		}
	}
	if sampleRate == 0 {
		sampleRate = 24000
	}
	if channels == 0 {
		channels = 1
	}
	return sampleRate, channels, nil
}

// GenerateSilence renders an mp3 of pure silence, format-matched so it can be
// concatenated with TTS segments before the final re-encode.
func GenerateSilence(durationSecs float64, sampleRate, channels int) ([]byte, error) {
	tmp := tmpFile("silence_")
	defer os.Remove(tmp)
	layout := "stereo"
	if channels == 1 {
		layout = "mono"
	}
	if _, err := run("ffmpeg",
		"-y", "-v", "error",
		"-f", "lavfi",
		"-i", fmt.Sprintf("anullsrc=r=%d:cl=%s", sampleRate, layout),
		"-t", strconv.FormatFloat(durationSecs, 'f', -1, 64),
		"-c:a", "libmp3lame", "-q:a", "2",
		tmp); err != nil {
		return nil, err
	}
	return os.ReadFile(tmp)
}

var outTimeRe = regexp.MustCompile(`out_time_us=(\d+)|out_time_ms=(\d+)`)

// DecodedDurationSecs is the true decoded duration in seconds: the number of
// samples the decoder actually emits, which is what concat and playback use.
// Differs from format=duration for files carrying a LAME/Xing gapless tag.
// Summed decoded durations equal the concat output exactly, so the read-along
// timeline stays locked to the audio.
func DecodedDurationSecs(file string) (float64, error) {
	out, err := run("ffmpeg", "-v", "error", "-i", file, "-f", "null", "-progress", "pipe:1", "-")
	if err != nil {
		return 0, err
	}
	matches := outTimeRe.FindAllString(out, -1)
	if len(matches) == 0 {
		return 0, fmt.Errorf("ffmpeg returned no duration for %s", file)
	}
	_, val, _ := strings.Cut(matches[len(matches)-1], "=")
	us, err := strconv.ParseInt(val, 10, 64)
	if err != nil || us <= 0 {
		return 0, fmt.Errorf("ffmpeg returned no duration for %s", file)
	}
	return float64(us) / 1e6, nil
}

func ProbeDurationSecs(file string) (float64, error) {
	out, err := run("ffprobe",
		"-v", "error",
		"-show_entries", "format=duration",
		"-of", "default=noprint_wrappers=1:nokey=1",
		file)
	if err != nil {
		return 0, err
	}
	secs, err := strconv.ParseFloat(strings.TrimSpace(out), 64)
	if err != nil || secs <= 0 {
		return 0, fmt.Errorf("ffprobe returned no duration for %s", file)
	}
	return secs, nil
}

// ProbeMp3Buffer measures the duration of an in-memory mp3 (for engines that
// don't return a duration header).
func ProbeMp3Buffer(buffer []byte) (float64, error) {
	tmp := tmpFile("probe_")
	defer os.Remove(tmp)
	if err := os.WriteFile(tmp, buffer, 0o644); err != nil {
		return 0, err
	}
	return ProbeDurationSecs(tmp)
}
